document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/transformer/MongoSelectionsUpdatingTransformation.java

```diff
@@ -0,0 +1,143 @@
+ *            "item": "$item",
+ *            "price": "$props.price"
+ *          },
+ *          "num_brands": {
+ *            "$distinct": "$brand"
+ *          }
+ *        }
+ *      },
+ *      {
+ *        "$project": {
+ *          "item": "$_id.item"
+ *        }
+ *      }
+ *   ]
+ * </code> since "item" appears in projection as well as grouping and "$distinctCount" is not
+ * supported
+ */
+final class MongoSelectionsUpdatingTransformation implements SelectingExpressionVisitor {
+  private static final Function<AggregateExpression, AggregateExpression> COUNT_HANDLER =
+      expression -> AggregateExpression.of(SUM, ConstantExpression.of(1));
+
+  private static final Function<AggregateExpression, AggregateExpression> DISTINCT_COUNT_HANDLER =
+      expression -> AggregateExpression.of(DISTINCT, expression.getExpression());
+
+  private static final Map<AggregationOperator, Function<AggregateExpression, AggregateExpression>>
+      AGGREGATION_SUBSTITUTE_MAP =
+          unmodifiableMap(
+              new EnumMap<>(AggregationOperator.class) {
+                {
+                  put(DISTINCT_COUNT, DISTINCT_COUNT_HANDLER);
+                  put(COUNT, COUNT_HANDLER);
+                }
+              });
+
+  private final List<GroupingExpression> aggregations;
+  private final SelectionSpec source;
+  private final Map<Integer, GroupingExpression> aggregationMap;
+
+  MongoSelectionsUpdatingTransformation(
+      List<GroupingExpression> aggregations, SelectionSpec source) {
+    this.aggregations = aggregations;
+    this.source = source;
+    this.aggregationMap = getAggregationMap();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final AggregateExpression expression) {
+    return substitute(expression);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final ConstantExpression expression) {
+    return source;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final FunctionExpression expression) {
+    return source;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public SelectionSpec visit(final IdentifierExpression expression) {
+    GroupingExpression matchingGroup = aggregationMap.get(expression.hashCode());
+    if (!expression.equals(matchingGroup)) {
+      return source;
+    }
+
+    String key = expression.getName();
+    String identifier = MongoCollection.ID_KEY + "." + key;
+
+    return SelectionSpec.of(IdentifierExpression.of(identifier), key);
+  }
+
+  private SelectionSpec substitute(final AggregateExpression expression) {
+    return Optional.ofNullable(AGGREGATION_SUBSTITUTE_MAP.get(expression.getAggregator()))
+        .map(newExp -> SelectionSpec.of(newExp.apply(expression), source.getAlias()))
+        .orElse(source);
+  }
+
+  private Map<Integer, GroupingExpression> getAggregationMap() {
+    return aggregations.stream().collect(toMap(GroupingExpression::hashCode, identity()));
```

Comment:
why do we need this map whose key is a hashcode?  How are we ensuring these grouping expressions don't get modified/transformed in the future? Seems error-prone.

Reply:
> why do we need this map whose key is a hashcode?
This is to identify a matching GroupingExpression in constant time. Without this, we will have to iterate through the list, which make this transformation take linear time.

> How are we ensuring these grouping expressions don't get modified/transformed in the future?
All the implementations of GroupingExpression are `@Value` classes (they are immutable). Hence, their hash code's can't change (during the lifecycle of the objects).

> Seems error-prone.
We are updating only the SelectingExpressions (or SelectionSpecs). None of the GroupingExpression are replaced/transformed in the query.

Reply:
I have renamed "aggregations" to "groupingExpressions" for improved clarity.

Reply:
As discussed, got rid of the map by foregoing optimization for better readability and robustness.